controllers/cluster/capabilities.go

```diff
@@ -0,0 +1,168 @@
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/pkg/errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+
+	v1 "github.com/openshift/api/config/v1"
+	configv1 "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+)
+
+var (
+	versionRegex = regexp.MustCompile(`(\d+)\.(\d+)`)
+)
+
+type Capabilities struct {
+	IsOnOpenshift, HasMachineAPI bool
+}
+
+type ServerVersion struct {
+	OcpVersion   string
+	Major, Minor int
+}
+
+func NewCapabilities(config *rest.Config) (Capabilities, error) {
+	var err error
+	var onOpenshift, machineAPI bool
+
+	if onOpenshift, err = isOnOpenshift(config); err != nil {
+		return Capabilities{}, errors.Wrap(err, "failed to check if cluster is on OpenShift")
+	}
+
+	if onOpenshift {
+		if machineAPI, err = hasMachineAPI(config); err != nil {
+			return Capabilities{}, errors.Wrap(err, "failed to check machine API capability")
+		}
+	}
+
+	return Capabilities{
+		IsOnOpenshift: onOpenshift,
+		HasMachineAPI: machineAPI,
+	}, nil
+}
+
+// isOnOpenshift returns true if the cluster has the openshift config group
+func isOnOpenshift(config *rest.Config) (bool, error) {
+	dc, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return false, err
+	}
+	apiGroups, err := dc.ServerGroups()
+	kind := schema.GroupVersionKind{Group: "config.openshift.io", Version: "v1", Kind: "ClusterVersion"}
+	for _, apiGroup := range apiGroups.Groups {
+		for _, supportedVersion := range apiGroup.Versions {
+			if supportedVersion.GroupVersion == kind.GroupVersion().String() {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+// hasMachineAPI returns true it the cluster has the MachineAPI
+func hasMachineAPI(config *rest.Config) (bool, error) {
+	client, err := configv1.NewForConfig(config)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to create cluster version client")
+	}
+	ocpVersion, err := getOpenshiftVersion(client)
+	if err != nil {
+		return false, errors.Wrap(err, "could not get OCP version")
+	}
+
+	if ocpVersion.Major == 4 && ocpVersion.Minor <= 13 {
```

Comment:
Nit: Technically we are still not treating correctly the use case of Major version < 4, I don't think it's an issue though, because of minKubeVersion - maybe just add a comment :thinking:

Reply:
> because of minKubeVersion

I didn't get this reference

> maybe just add a comment

I will change the major version check